lib/spack/spack/filesystem_view.py

```diff
@@ -416,6 +416,18 @@ def remove_specs(self, *specs, **kwargs):
         remove_extension = ft.partial(self.remove_extension,
                                       with_dependents=with_dependents)
 
+        tty.debug('remove extensions %s' % [s.name for s in extensions])
+
+        extensions_sorted = []
+        for s in extensions:
+            if any(t in extensions_sorted for t in s.dependencies()):
```

Comment:
Hi @js947 this is an interesting approach but I think it may run into problems because `Spec.dependencies() only returns immediate dependencies. So for example if you have:

```
W->X->Y->Z
 \______/
```

And the arbitrarily-ordered `extensions` look like `[Y, W, X, Z]`.

There is likely a way to update this approach to work although I think a simpler sorting algorithm might be to sort the specs based on their total size (`len(spec.traverse())`) - if `X->Y` then `X` will be larger. Would you be interested/able to try that approach?

If so, one additional minor detail: this will have to set `deptype='all'` when invoking `spec.traverse` like `spec.traverse(deptype='all')`. That might be the default but if it ever changes it could interfere with the size calculations (which I can add more detail on if you're interested in pursuing it).

Reply:
Hi @scheibelp, thanks for the suggestion. I don't quite follow what you mean about the length of the spec, but I used the traverse method in an improved implementation of a topological sort.